File: java/src/bfs/randomTestData.java
package bfs;

import java.util.*;

class randomTestData {
	private static final Random rand = new Random(0);

	List<List<Integer>> graph;
	int start, end;
	int minDist;
	List<Integer> minPath = new LinkedList<>();
	List<Integer> order = new ArrayList<>();

	randomTestData() {
		int n = 1 + rand.nextInt(50), maxAdj = rand.nextInt(50);
		graph = new ArrayList<>();
		for (var u = 0; u < n; u++) {
			var adj = new ArrayList<Integer>();

			var m = rand.nextInt(1 + maxAdj);
			for (var i = 0; i < m; i++) {
				int v = rand.nextInt(n);
				adj.add(v);
			}

			graph.add(adj);
		}

		start = rand.nextInt(n);
		end = rand.nextInt(n);
		walk();
		minDist = minPath.size() - 1;
	}

	private void walk() {
		var prev = new int[graph.size()];
		Arrays.fill(prev, -1);

		order.addLast(start);
		prev[start] = start;

		for (int i = 0; i < order.size(); i++) {
			var u = order.get(i);
			for (var v: graph.get(u)) {
				if (prev[v] < 0) {
					order.addLast(v);
					prev[v] = u;
				}
			}
		}

		if (prev[end] < 0) {
			return;
		}

		for (var v = end;; v = prev[v]) {
			minPath.addFirst(v);
			if (v == prev[v]) break;
		}
	}
}

File: go/bfs/random_test_data.go
package bfs

import "math/rand/v2"

type randomTestData struct {
	graph               [][]int
	start, end, minDist int
	minPath, order      []int
}

func newRandomTestData() *randomTestData {
	rand := rand.New(rand.NewPCG(0, 0))
	n := 1 + rand.IntN(50)
	maxAdj := rand.IntN(50)

	td := &randomTestData{}
	for range n {
		var adj []int
		m := rand.IntN(1 + maxAdj)
		for range m {
			v := rand.IntN(n)
			adj = append(adj, v)
		}

		td.graph = append(td.graph, adj)
	}

	td.start = rand.IntN(n)
	td.end = rand.IntN(n)
	td.walk()
	td.minDist = len(td.minPath) - 1

	return td
}

func (td *randomTestData) walk() {
	prev := make([]int, len(td.graph))
	for i := range prev {
		prev[i] = -1
	}

	td.order = append(td.order, td.start)
	prev[td.start] = td.start

	for i := 0; i < len(td.order); i++ {
		u := td.order[i]
		for _, v := range td.graph[u] {
			if prev[v] < 0 {
				td.order = append(td.order, v)
				prev[v] = u
			}
		}
	}

	if prev[td.end] < 0 {
		return
	}

	for v := td.end; ; v = prev[v] {
		td.minPath = append(td.minPath, v)
		if v == prev[v] {
			break
		}
	}

	for l, r := 0, len(td.minPath); l < r; l, r = l+1, r-1 {
		td.minPath[l], td.minPath[r] = td.minPath[r], td.minPath[l]
	}
}
